SharedUtils/src/main/java/com/braintreepayments/api/sharedutils/HttpClient.kt

```diff
@@ -0,0 +1,110 @@
+package com.braintreepayments.api.sharedutils
+
+import androidx.annotation.RestrictTo
+import java.net.MalformedURLException
+import java.net.URISyntaxException
+import java.net.URL
+import javax.net.ssl.SSLSocketFactory
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+class HttpClient internal constructor(
+    private val syncHttpClient: SynchronousHttpClient,
+    private val scheduler: Scheduler
+) {
+    enum class RetryStrategy { NO_RETRY, RETRY_MAX_3_TIMES }
+
+    private val retryCountMap: MutableMap<URL, Int> = HashMap()
+
+    constructor(
+        socketFactory: SSLSocketFactory,
+        httpResponseParser: HttpResponseParser
+    ) : this(
+        syncHttpClient = SynchronousHttpClient(socketFactory, httpResponseParser),
+        scheduler = ThreadScheduler()
+    )
+
+    @Throws(Exception::class)
+    fun sendRequest(request: HttpRequest): String {
+        return syncHttpClient.request(request).body ?: ""
+    }
+
+    fun sendRequest(
+        request: HttpRequest,
+        callback: NetworkResponseCallback?,
+        retryStrategy: RetryStrategy = RetryStrategy.NO_RETRY,
+    ) {
+        scheduleRequest(request, retryStrategy, callback)
+    }
+
+    @Suppress("TooGenericExceptionCaught")
+    private fun scheduleRequest(
+        request: HttpRequest,
+        retryStrategy: RetryStrategy,
+        callback: NetworkResponseCallback?
+    ) {
+        resetRetryCount(request)
+
+        scheduler.runOnBackground {
+            try {
+                val httpResponse = syncHttpClient.request(request)
+                callback?.let {
+                    scheduler.runOnMain { callback.onResult(httpResponse, null) }
+                }
+            } catch (e: Exception) {
+                when (retryStrategy) {
+                    RetryStrategy.NO_RETRY -> notifyErrorOnMainThread(callback, e)
+                    RetryStrategy.RETRY_MAX_3_TIMES -> retryGet(request, retryStrategy, callback)
+                }
+            }
+        }
+    }
+
+    private fun retryGet(
+        request: HttpRequest,
+        retryStrategy: RetryStrategy,
+        callback: NetworkResponseCallback?
+    ) {
+        var url: URL? = null
+        try {
+            url = request.url
+        } catch (ignore: MalformedURLException) {
+        } catch (ignore: URISyntaxException) {
+        }
```

Comment:
Are we able to add a comment or something in there blocks to indicate why we ignore these? To me they seem like something we wouldn't want to ignore, but I see we were also doing this previously.

Reply:
I'm not too sure why we're catching these 😅 but from a quick look, `HttpRequest.getURL()` can throw these exceptions and it feels like we don't want to crash due to those.

I think a lot of the networking code can be refactored to remove the ambiguity, but I wanted to keep the actual code changes to a minimum other than the java to kotlin conversion.

Reply:
That sounds good, yeah it's not blocking was more for my general understanding.